src/realm.js

```diff
@@ -683,19 +683,37 @@ export class Realm {
       value: void | Value
     ) => void
   ) {
-    let saved_createdObjectsTrackedForLeaks = this.createdObjectsTrackedForLeaks;
     let saved_reportSideEffectCallback = this.reportSideEffectCallback;
-    // Track all objects (including function closures) created during
-    // this call. This will be used to make the assumption that every
-    // *other* object is unchanged (pure). These objects are marked
-    // as leaked if they're passed to abstract functions.
-    this.createdObjectsTrackedForLeaks = new Set();
-    this.reportSideEffectCallback = reportSideEffectFunc;
+    let shouldClearCreatedObjectsTrackedForLeaks = false;
+    // Only create a tracked objects set if we haven't already got one
+    // otherwise, we'll end up creating a new set for nested evalautePure
+    // calls.
+    if (this.createdObjectsTrackedForLeaks === undefined) {
```

Comment:
I'm not sure this makes sense.

With this code, every time we enter a "nested" pure scope, we're reusing the same set of tracked objects. So we're effectively having them "share" the same top-level pure scope.

If my understanding is correct, every pure scope should have *its own set of tracked objects* associated with it. When we enter a nested pure scope, we should create a new set. When we leave it, we should restore the previous set.

I can’t immediately give you an example of why this matters, but conceptually that approach would make more sense to me. Are there any reasons why you didn’t implement it that way? I can try to come up with a specific example if this explanation doesn’t make sense.

Reply:
I just realized that what I'm describing is pretty much what the original code was doing.

Why did this part need to change? I thought you're fixing the side effect detection part.

Reply:
What do nested pure scopes even mean?

Normally, a pure scope means that it's safe to make abstract calls inside it because we know they won't touch anything except what we created in it.

```js
var outer = {};
__evaluatePureFunction(() => {
  var x = {};
  // we know `outer` won't change so we don't havoc it:
  abstract(outer);
  // this will havoc `x`:
  abstract(outer, x); 
});
```

Here’s the scenario I’m worried about:

```js
var outer = {};
__evaluatePureFunction(() => {
  var x = {};
  __evaluatePureFunction(() => {
    var y = {};

    // we know `outer` won't change so we don't havoc it:
    abstract(outer);

    // should this havoc `x` or not?
    // my intuition is that it *shouldn't* because the inner scope is "pure".
    // but I think this PR would change that.
    abstract(x);

    // this will havoc `y`:
    abstract(y);
  });
});
```

Reply:
I'll take another look at this and see why again.

Reply:
I looked into this more, and the reason I did this was because of `this`. Take

```
class Object {
  doSomething() {
    this.x = 0;
    __evaluatePureFunction(() => { this.x++ } );
  }
}
```

Mutating `this.x` should not fail, it's still mutating the same `this` because of lexical scoping. It currently does fail, which is not correct. So what I did was make the objects created in parent pure scopes pass into children scopes,  thus correcting this issue (but as you mentioned above, breaking other things). It's a tricky one, these issues only come into play when using this PR with https://github.com/facebook/prepack/pull/2165 - as it introduces a pureScope for every single function call, meaning there are hundreds if not thousands of pure scopes, vs like a few like we have now.

Reply:
I'll take another look into this tomorrow, but for now I've reverted the change and put in the original, to at least unblock this PR.

Reply:
@gaearon Please, if you have time, could you add some tests cases around this (as generic serializer) tests like you did above. The above case didn't seem to fail for me with the previous changes, so I'm wondering if something else is not right.